### PR TITLE
Particle system

### PR DESCRIPTION
**Pull request by iamgreaser**
Will still need a lot of work. Posting this PR right now mostly because I'm trying to suss out where everything is supposed to even go.

### CODE REVIEW

File: mappings/net/minecraft/client/particle/Particle.mapping [hunk 1]
```diff
@@ -0,0 +1,81 @@
+CLASS none/bnl net/minecraft/client/particle/Particle
+	FIELD A colorRed F
+	FIELD B colorGreen F
+	FIELD C colorBlue F
+	FIELD D colorAlpha F
+	FIELD E sprite Lnone/bxe;
```
Comment (copygirl) [inline mappings/net/minecraft/client/particle/Particle.mapping:6]: We haven't used sprite before, why start now?

File: mappings/net/minecraft/client/particle/ParticleManager.mapping [hunk 2]
```diff
@@ -0,0 +1,37 @@
+CLASS none/bnm net/minecraft/client/particle/ParticleManager
+	FIELD a world Lnone/aiw;
+	FIELD b PARTICLE_TEX Lnone/kp;
+	FIELD c particleQueues [[Ljava/util/ArrayDeque;
+	FIELD e textureManager Lnone/bxf;
+	FIELD f random Ljava/util/Random;
```
Comment (copygirl) [inline mappings/net/minecraft/client/particle/ParticleManager.mapping]: `rand` (This isn't static? Odd.)

File: mappings/net/minecraft/client/particle/ParticleManager.mapping [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+		ARG 2 y
+		ARG 3 z
+		ARG 4 velocityX
+		ARG 5 velocityY
+		ARG 6 velocityZ
+		ARG 7 extra
+	METHOD a registerFactory (ILnone/bnn;)V
+		ARG 0 particleId
+		ARG 1 factory
+	METHOD a tickParticleQueue (Ljava/util/Queue;)V
```
Comment (copygirl) [inline mappings/net/minecraft/client/particle/ParticleManager.mapping]: `update`?

File: mappings/net/minecraft/client/particle/ParticleManager.mapping [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+		ARG 0 particleQueue
+	METHOD a setWorld (Lnone/aiw;)V
+		ARG 0 world
+	METHOD a addParticle (Lnone/bnl;)V
+		ARG 0 particle
+	METHOD b getParticleCountAsString ()Ljava/lang/String;
+	METHOD b tickParticle (Lnone/bnl;)V
```
Comment (copygirl) [inline mappings/net/minecraft/client/particle/ParticleManager.mapping]: `update`?

File: mappings/net/minecraft/client/particle/Particle.mapping [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+		ARG 3 z
+	METHOD <init> (Lnone/aiw;DDDDDD)V
+		ARG 0 world
+		ARG 1 x
+		ARG 2 y
+		ARG 3 z
+		ARG 4 velocityX
+		ARG 5 velocityY
+		ARG 6 velocityZ
+	METHOD a tick ()V
```
Comment (copygirl) [inline mappings/net/minecraft/client/particle/Particle.mapping]: `update`? We should keep things consistent and I believe we've been using `update` for everything.

**Review by copygirl — CHANGES_REQUESTED** [review]

File: mappings/net/minecraft/client/particle/Particle.mapping (hunk 5 above)
Comment (iamgreaser) [reply mappings/net/minecraft/client/particle/Particle.mapping]: `tick` is used in other places.

File: mappings/net/minecraft/client/particle/Particle.mapping (hunk 1 above)
Comment (iamgreaser) [reply mappings/net/minecraft/client/particle/Particle.mapping:6]: There's a class called `Sprite`, what are these instances _usually_ called?

File: mappings/net/minecraft/client/particle/Particle.mapping (hunk 5 above)
Comment (iamgreaser) [reply mappings/net/minecraft/client/particle/Particle.mapping]: Also, `tick` is also less vague. If you insist though I can rename it to update.

File: mappings/net/minecraft/client/particle/ParticleManager.mapping (hunk 3 above)
Comment (iamgreaser) [reply mappings/net/minecraft/client/particle/ParticleManager.mapping]: Comment as per `tick`->`update` but I will NOT overload these into all being called `update`.

Comment (exaskye) [reply mappings/net/minecraft/client/particle/ParticleManager.mapping]: I think they mean `updateParticleQueue`, not just `update`.

File: mappings/net/minecraft/client/particle/Particle.mapping (hunk 1 above)
Comment (exaskye) [reply mappings/net/minecraft/client/particle/Particle.mapping:6]: none/bxe _is_ `Sprite`, so this seems fine to me.

File: mappings/net/minecraft/client/particle/Particle.mapping (hunk 5 above)
Comment (copygirl) [reply mappings/net/minecraft/client/particle/Particle.mapping]: I agree. Then we should update all other mentions of update, too.

**Comment by iamgreaser** [comment]
OK, how should all the different particles be named?

Should we go e.g. `ParticleFire` or `FireParticle`? The former is hell from a tab-completion perspective, and they're all going to be in `net.minecraft.client.particle.*` anyway.

**Comment by darkevilmac** [comment]
> OK, how should all the different particles be named?
> 
> Should we go e.g. ParticleFire or FireParticle? The former is hell from a tab-completion perspective, and they're all going to be in net.minecraft.client.particle.\* anyway.

ParticleFire matches better to the current naming of other things.

**Comment by iamgreaser** [comment]
I guess from that perspective you won't be using the classes directly, so I guess `ParticleFire` could still work fine.

**Review by copygirl — APPROVED** [review]

**Review by exaskye — APPROVED** [review]

**Comment by iamgreaser** [comment]
Would you rather I dropped the WIP tag and had this pulled now, or would you rather I get names for the particle classes first?

**Comment by copygirl** [comment]
This PR looks good to me so I can merge it and have another one for the particle class names.

**Comment by iamgreaser** [comment]
@copygirl Go for it.

**Pull request merged**